python/pylibcudf/pylibcudf/interop.pyx

```diff
@@ -252,82 +188,25 @@ def _to_arrow_datatype(cudf_object, **kwargs):
-    )
-    release_arrow_schema_raw(schema)
-
-
-cdef void _release_array(object array_capsule) noexcept:
-    """Release the ArrowArray object stored in a PyCapsule."""
-    cdef ArrowArray* array = <ArrowArray*>PyCapsule_GetPointer(
-        array_capsule, 'arrow_array'
-    )
-    release_arrow_array_raw(array)
-
-
-def _maybe_create_nested_column_metadata(Column col):
-    return ColumnMetadata(
-        children_meta=[
-            _maybe_create_nested_column_metadata(child) for child in col.children()
-        ]
-    )
-
-
-def _table_to_schema(Table tbl, metadata):
-    if metadata is None:
-        metadata = [_maybe_create_nested_column_metadata(col) for col in tbl.columns()]
-    else:
-        metadata = [ColumnMetadata(m) if isinstance(m, str) else m for m in metadata]
-
-    cdef vector[column_metadata] c_metadata
-    c_metadata.reserve(len(metadata))
-    for meta in metadata:
-        c_metadata.push_back(_metadata_to_libcudf(meta))
-
-    cdef ArrowSchema* raw_schema_ptr
-    with nogil:
-        raw_schema_ptr = to_arrow_schema_raw(tbl.view(), c_metadata)
-
-    return PyCapsule_New(<void*>raw_schema_ptr, 'arrow_schema', _release_schema)
-
-
-def _table_to_host_array(Table tbl):
-    cdef ArrowArray* raw_host_array_ptr
-    with nogil:
-        raw_host_array_ptr = to_arrow_host_raw(tbl.view())
-
-    return PyCapsule_New(<void*>raw_host_array_ptr, "arrow_array", _release_array)
-
-
-class _TableWithArrowMetadata:
-    def __init__(self, tbl, metadata=None):
-        self.tbl = tbl
+class _ObjectWithArrowMetadata:
+    def __init__(self, obj, metadata=None):
+        self.obj = obj
         self.metadata = metadata
 
     def __arrow_c_array__(self, requested_schema=None):
-        return _table_to_schema(self.tbl, self.metadata), _table_to_host_array(self.tbl)
+        return self.obj._to_schema(self.metadata), self.obj._to_host_array()
 
 
-# TODO: In the long run we should get rid of the `to_arrow` functions in favor of using
-# the protocols directly via `pa.table(cudf_object, schema=...)` directly. We can do the
-# same for columns. We cannot do this for scalars since there is no corresponding
-# protocol. Since this will require broader changes throughout the codebase, the current
-# approach is to leverage the protocol internally but to continue exposing `to_arrow`.
 @to_arrow.register(Table)
 def _to_arrow_table(cudf_object, metadata=None):
-    test_table = _TableWithArrowMetadata(cudf_object, metadata)
-    return pa.table(test_table)
+    """Create a PyArrow table from a pylibcudf table."""
+    return pa.table(_ObjectWithArrowMetadata(cudf_object, metadata))
 
 
 @to_arrow.register(Column)
 def _to_arrow_array(cudf_object, metadata=None):
     """Create a PyArrow array from a pylibcudf column."""
-    if metadata is not None:
-        metadata = [metadata]
-    return to_arrow(Table([cudf_object]), metadata)[0]
+    return pa.array(_ObjectWithArrowMetadata(cudf_object, metadata))
```

Comment:
```suggestion
    return pa.array(_ObjectWithArrowMetadata(plc_obj, metadata))
```

Reply:
Done in https://github.com/rapidsai/cudf/pull/18402/commits/fd5ab31a9e5cd432ffc01f30b98878278bdb0c5e